### PR TITLE
New registration method for simulation

### PR DESCRIPTION
**Pull request by mpardalos**
# Summary

The visualisation team has decided to change the way clients are registered to make it easier to have multiple configurations for every client. We also want to allow for running simulations of every island being controlled by the same agent (so e.g. we have 6 instances of team 3). This means that the **team id is no longer constant**. I have updated each teams agent to reflect that but that means I need approval from every team.

## Test Plan

All existing tests pass, and the only code added to the server is "configuration code" so it should not need testing.

### CODE REVIEW

**Comment by mpardalos** [comment]
CC @Eirikalb, @ezgimez, @Spruudis, @JamesTev, @P-A-Ha, @kunjzk, @justma360

**Comment by mpardalos** [comment]
What does `File is not goimports-ed` mean? @lhl2617

**Comment by lhl2617** [comment]
> What does `File is not goimports-ed` mean? @lhl2617

@mpardalos  the sequence of the imports is alphabetical internal libs, space, then alphabetical extern libs.

**Review by Eirikalb — APPROVED** [review]
I mike we trust

**Comment by JamesTev** [comment]
Looks good - at a glance at least. If it still runs and passes tests, I trust you and am happy for you to go ahead.

File: internal/server/server.go
Comment (lhl2617) [inline internal/server/server.go]: you want a `Factory` method for each team. so that the client pointers containing the state are not preserved between consecutive runs (i.e. each new run has a new clean state)

this was the cause of the annoying `NaN` bug on wasm (because we run repeat sims on the same go instances and same pointers across runs)

**Review by JamesTev — APPROVED** [review]
Approving but obviously wait for feedback from other teams before merging.

File: internal/server/server.go
Comment (lhl2617) [reply internal/server/server.go]: So instead you should write `DefaultClientFactory`

**Comment by mpardalos** [comment]
> Approving but obviously wait for feedback from other teams before merging.

Thanks. Will do.

**Review by Neelesh99 — APPROVED** [review]
LGTM

**Review by lhl2617 — CHANGES_REQUESTED** [review]
Factory model required to return new pointer every run

**Review by lhl2617 — APPROVED** [review]
ok boi

File: internal/server/server.go
Comment (mpardalos) [reply internal/server/server.go]: Resolved on discord

**Comment by mpardalos** [comment]
What team is still missing for approval?

**Comment by NGovani** [comment]
> What team is still missing for approval?

Team 6, you forgot to tag @walkeronwater 
I assume @kunjzk laughing at your initial tag was confirmation

EDIT: I have no clue who is charge of team 6 agent could be @Diregie-J or @Regnigh

**Review by Ning-72 — APPROVED** [review]

File: internal/clients/team3/client.go
Comment (lhl2617) [inline internal/clients/team3/client.go]: remove

File: internal/clients/team4/client.go
Comment (lhl2617) [inline internal/clients/team4/client.go]: remove

**Comment by walkeronwater** [comment]
> > What team is still missing for approval?
> 
> Team 6, you forgot to tag @walkeronwater
> I assume @kunjzk laughing at your initial tag was confirmation
> 
> EDIT: I have no clue who is charge of team 6 agent could be @Diregie-J or @Regnigh

#404 not found - @Diregie-J will approve

**Comment by mpardalos** [comment]
In addition to team 6: Team 3 tests break. Idk who to tag from team 3. I'd appreciate it if @NGovani could tag the right person.

EDIT: @Neelesh99 you're team 3 right? Can you tell me how to fix the tests/fix them yourself?

File: internal/clients/team4/client.go
Comment (mpardalos) [reply internal/clients/team4/client.go]: Done

File: internal/clients/team3/client.go
Comment (mpardalos) [reply internal/clients/team3/client.go]: Did it. Broke team 3 tests. This code hates me.

**Comment by NGovani** [comment]
> In addition to team 6: Team 3 tests break. Idk who to tag from team 3. I'd appreciate it if @NGovani could tag the right person.
> 
> EDIT: @Neelesh99 you're team 3 right? Can you tell me how to fix the tests/fix them yourself?

Should be an easy fix
https://github.com/SOMAS2020/SOMAS2020/blob/32a401171c5b82b1ce199c10e5b3cece5dadb01d/internal/clients/team3/clientiitoiifo_test.go#L21
here just make the id = shared.Team3 for the base client struct
same for all testcases, they give correct input so their maps dont have their own ID in them. which fucks over the tests that assumes id is constantly 3

**Comment by mpardalos** [comment]
@lhl2617 CI is complaining for every team's code cause I've touched everything. Ignore and merge?

**Pull request merged**
